### ISSUE TITLE
Feature: add access_modes to CRD provisionercapabilities.storage.kubesphere.io

### ISSUE THREAD

**Issue by stoneshi-yunify** [kubesphere/kubesphere#3085]
When creating a PVC/volume in ks console, we want the user to select the access mode for this volume, see below screenshot.

![image](https://user-images.githubusercontent.com/70880165/98195249-387b7780-1f5c-11eb-9c7f-c47e012a1b57.png)

We want to only display the available access modes that the user can choose, so that the user won't make a mistake.

The available access modes come from the storageclass's annotation:
```
$ kubectl get sc csi-neonsan -o yaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    meta.helm.sh/release-name: csi-neonsan
    meta.helm.sh/release-namespace: kube-system
    storageclass.kubesphere.io/support-snapshot: "true"
    storageclass.kubesphere.io/supported-access-modes: '["ReadWriteOnce"]'   <-----
  creationTimestamp: "2020-09-22T09:44:40Z"
```

And this annotation is added when creates the storage class, he can choose what access modes this provisioner supports:
![image](https://user-images.githubusercontent.com/70880165/98350420-46adbe80-2056-11eb-9dca-87153a8b4d78.png)

But we want to automate this process, we want to probe the access modes automatically for the provisioner, because sometimes the user does not exactly know what access modes this provisioner supports.

We have two CRDs that might be helpful to this issue. They are:
```
$ kubectl get crds | grep cap
provisionercapabilities.storage.kubesphere.io                2020-09-22T02:46:56Z
storageclasscapabilities.storage.kubesphere.io               2020-09-22T02:46:56Z
```

The `provisionercapabilities` contains the capabilities for the provisioner (the storage driver), and the `storageclasscapabilities` just borrows the capabilities from the provisioner. Each storage class links to a storageclasscapabilities.

So the thought is adding the access_modes into provisionercapabilities, so that the storageclasscapabilities will have it too. Then when user creates volume in ks console, we query the corresponding storageclasscapabilities for selected storageclass, and display available access modes for this volume.

```
root@i-os8hfxab:~
$ kubectl get provisionercapabilities.storage.kubesphere.io -o yaml
apiVersion: v1
items:
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: ceph-com-rbd
    resourceVersion: "2207"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/ceph-com-rbd
    uid: 0d4cbc60-4a91-4877-91d0-e5104dd27136
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: UNKNOWN
        list: false
        stats: false
    pluginInfo:
      name: ceph.com/rbd
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-aws-ebs
    resourceVersion: "2200"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-aws-ebs
    uid: d75d80d6-441f-4eed-9354-01f12ae89963
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: ONLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/aws-ebs
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-azure-disk
    resourceVersion: "2201"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-azure-disk
    uid: d0f787fc-ab0e-44f2-a75e-e656a3800fc6
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: OFFLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/azure-disk
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-azure-file
    resourceVersion: "2202"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-azure-file
    uid: f9fc6097-b757-481b-afba-f268a3c89a7c
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: OFFLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/azure-file
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-cinder
    resourceVersion: "2203"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-cinder
    uid: 2e9f64a2-4643-4e32-96f9-3aea0ccf286b
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: ONLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/cinder
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-gce-pd
    resourceVersion: "2204"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-gce-pd
    uid: c962e228-6d8a-4d4d-9c8c-5cb25178fcc4
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: ONLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/gce-pd
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-glusterfs
    resourceVersion: "2205"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-glusterfs
    uid: bb8e8508-309e-4f54-b415-0795cd615ac9
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: OFFLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/glusterfs
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-portworx-volume
    resourceVersion: "2206"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-portworx-volume
    uid: 6634ab00-2089-4786-aff3-47dfdf5c3846
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: OFFLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/portworx-volume
      version: ""
- apiVersion: storage.kubesphere.io/v1alpha1
  kind: ProvisionerCapability
  metadata:
    creationTimestamp: "2020-09-22T02:46:57Z"
    generation: 1
    name: kubernetes-io-rbd
    resourceVersion: "2208"
    selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/kubernetes-io-rbd
    uid: 045436c3-5bd6-4fbf-88ad-3cc08ec0f86a
  spec:
    features:
      snapshot:
        create: false
        list: false
      topology: false
      volume:
        attach: true
        clone: false
        create: true
        expandMode: ONLINE
        list: false
        stats: false
    pluginInfo:
      name: kubernetes.io/rbd
      version: ""
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
root@i-os8hfxab:~
$ kubectl get provisionercapabilities.storage.kubesphere.io
NAME                            PROVISIONER                     EXPAND    AGE
ceph-com-rbd                    ceph.com/rbd                    UNKNOWN   44d
kubernetes-io-aws-ebs           kubernetes.io/aws-ebs           ONLINE    44d
kubernetes-io-azure-disk        kubernetes.io/azure-disk        OFFLINE   44d
kubernetes-io-azure-file        kubernetes.io/azure-file        OFFLINE   44d
kubernetes-io-cinder            kubernetes.io/cinder            ONLINE    44d
kubernetes-io-gce-pd            kubernetes.io/gce-pd            ONLINE    44d
kubernetes-io-glusterfs         kubernetes.io/glusterfs         OFFLINE   44d
kubernetes-io-portworx-volume   kubernetes.io/portworx-volume   OFFLINE   44d
kubernetes-io-rbd               kubernetes.io/rbd               ONLINE    44d
root@i-os8hfxab:~
$ kubectl get provisionercapabilities.storage.kubesphere.io ceph-com-rbd -o yaml
apiVersion: storage.kubesphere.io/v1alpha1
kind: ProvisionerCapability
metadata:
  creationTimestamp: "2020-09-22T02:46:57Z"
  generation: 1
  name: ceph-com-rbd
  resourceVersion: "2207"
  selfLink: /apis/storage.kubesphere.io/v1alpha1/provisionercapabilities/ceph-com-rbd
  uid: 0d4cbc60-4a91-4877-91d0-e5104dd27136
spec:
  features:
    snapshot:
      create: false
      list: false
    topology: false
    volume:
      attach: true
      clone: false
      create: true
      expandMode: UNKNOWN
      list: false
      stats: false
  pluginInfo:
    name: ceph.com/rbd
    version: ""
root@i-os8hfxab:~
$
root@i-os8hfxab:~
$
root@i-os8hfxab:~
$ kubectl get storageclasscapabilities.storage.kubesphere.io
NAME            PROVISIONER                VOLUME   EXPAND    CLONE   SNAPSHOT   AGE
csi-neonsan     neonsan.csi.qingstor.com   true     OFFLINE   true    true       43d
csi-qingcloud   disk.csi.qingcloud.com     true     OFFLINE   true    true       43d
glusterfs       kubernetes.io/glusterfs    true     OFFLINE   false   false      43d
rbd             ceph.com/rbd               true     UNKNOWN   false   false      43d
root@i-os8hfxab:~
$  kubectl get storageclasscapabilities.storage.kubesphere.io rbd -o yaml
apiVersion: storage.kubesphere.io/v1alpha1
kind: StorageClassCapability
metadata:
  creationTimestamp: "2020-09-23T01:53:30Z"
  generation: 1
  name: rbd
  resourceVersion: "456069"
  selfLink: /apis/storage.kubesphere.io/v1alpha1/storageclasscapabilities/rbd
  uid: 126c371e-b4bd-48e7-9fed-236a409a9170
spec:
  features:
    snapshot:
      create: false
      list: false
    topology: false
    volume:
      attach: true
      clone: false
      create: true
      expandMode: UNKNOWN
      list: false
      stats: false
  provisioner: ceph.com/rbd
root@i-os8hfxab:~
$
root@i-os8hfxab:~
$
root@i-os8hfxab:~
$ kubectl get sc
NAME              PROVISIONER                            RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
asdfasd           adsfaasd                               Delete          Immediate              false                  25h
csi-neonsan       neonsan.csi.qingstor.com               Delete          WaitForFirstConsumer   true                   43d
csi-qingcloud     disk.csi.qingcloud.com                 Delete          WaitForFirstConsumer   true                   43d
csi-rbd-sc        rbd.csi.ceph.com                       Delete          Immediate              true                   43d
glusterfs         kubernetes.io/glusterfs                Delete          Immediate              true                   43d
local (default)   openebs.io/local                       Delete          WaitForFirstConsumer   false                  44d
nfs-client        cluster.local/nfs-client-provisioner   Delete          Immediate              true                   43d
rbd               ceph.com/rbd                           Delete          Immediate              false                  43d
test              yunify.com/stone                       Delete          Immediate              true                   25h
root@i-os8hfxab:~
$ kubectl get sc rbd -o yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    meta.helm.sh/release-name: rbd-provisioner
    meta.helm.sh/release-namespace: kube-system
    storageclass.kubesphere.io/support-snapshot: "false"
  creationTimestamp: "2020-09-23T01:53:30Z"
  labels:
    app.kubernetes.io/managed-by: Helm
  name: rbd
  resourceVersion: "456067"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/rbd
  uid: ae5a3e94-608e-458e-8e2d-426beb0383e1
parameters:
  adminId: admin
  adminSecretName: ceph-secret-admin
  adminSecretNamespace: kube-system
  fsType: ext4
  imageFeatures: layering
  imageFormat: "2"
  monitors: 192.168.0.12:6789
  pool: rbd
  userId: admin
  userSecretName: ceph-secret
  userSecretNamespace: kube-system
provisioner: ceph.com/rbd
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

**Comment by stoneshi-yunify** [comment]
We have installed some provisionercapabilities for some well-known provisioners when installing KS. Obviously this can not cover all storage systems, so how we populate the storageclasscapabilities for those provisioners not covered?

If it's a CSI provisioner, we can use the CSI RPC call to query the CSI controller capabilities, CSI node capabilities, and the CSI plugin capabilities, then fill the info into the storageclasscapabilities.

If it isn't a CSI provisioner, nowadays this is rare case, we have no way to get the access_modes info. The storage/KS admin will have to create a provisionercapabilities object on cluster manually.

Even it's a CSI provisioner, the CSI RPC call method still have some problems, the ks-controller-manager might not be able to connect to CSI controller/node because they might not be running on the same node, or the unix domain socket the CSI containers are using is not accessible outside of the CSI pod.

**Comment by stoneshi-yunify** [comment]
It looks like there isn't a reliable way to get storage provider's capabilities (expand, clone, snapshot etc) automatically, as the reason I stated in comment above. So we have to ask the user to input/set the capabilities by self.

I also checked rancher and openshift, none of them have such feature. They don't even support expand/clone/snapshot volume in UI. For the access modes, they will ask user to select the access modes when creating the pvc.

We may defer this issue to v4.0 as we're running out of time for v3.1.
